Src/IronPython/Compiler/Parser.cs

```diff
@@ -1696,14 +1696,17 @@ struct WithItem {
                     return ErrorStmt();
                 }
 
-                if (metaclass != null) {
-                    l.Add(new AssignmentStatement(new[] { new NameExpression("__metaclass__") }, metaclass));
-                }
-
+                bool addMetaclass = true;
                 while (true) {
                     Statement s = ParseStmt();
 
                     l.Add(s);
+
+                    if(addMetaclass && metaclass != null) {
```

Comment:
If you wanted to be minimal about it, you could probably get rid of `addMetaclass` and just move it after the while loop with an `l.Insert(1, ...)`.